rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -108,66 +134,228 @@ public void close(ISourceLocation loc) {
         }
     }
 
+    /**
+     * This class keeps together diagnostics messages and the versioned AST for
+     * which the diagnostics were computed. This is needed to ensure that
+     * builder diagnostics are calculated for the same AST as
+     * supposedly-corresponding analyzer diagnostics.
+     */
+    private static class VersionedDiagnostics {
+        public final VersionedTree tree;
+        public final List<Diagnostic> messages;
+
+        public VersionedDiagnostics(VersionedTree tree, List<Diagnostic> messages) {
+            this.tree = tree;
+            this.messages = messages;
+        }
+
+        public static boolean setIfNewer(AtomicReference<VersionedDiagnostics> box, VersionedDiagnostics maybeNewer) {
```

Comment:
maybe the whole function can get a better name? something like `replaceIfNewer`?